packages/gatsby/src/datastore/common/iterable.ts

```diff
@@ -1,44 +1,153 @@
+    const source =
+      typeof this.source === `function` ? this.source() : this.source
+    return source[Symbol.iterator]()
   }
 
-  concat<U>(other: Iterable<U>): GatsbyIterable<T | U> {
-    return new GatsbyIterable(concatSequence(this, other))
+  concat<U = T>(other: Iterable<U>): GatsbyIterable<T | U> {
+    return new GatsbyIterable(() => concatSequence(this, other))
   }
 
-  map<U>(fn: (entry: T) => U): GatsbyIterable<U> {
-    return new GatsbyIterable(mapSequence(this, fn))
+  map<U>(fn: (entry: T, index: number) => U): GatsbyIterable<U> {
+    return new GatsbyIterable(() => mapSequence(this, fn))
   }
 
   filter(predicate: (entry: T) => unknown): GatsbyIterable<T> {
-    return new GatsbyIterable<T>(filterSequence(this, predicate))
+    return new GatsbyIterable(() => filterSequence(this, predicate))
+  }
+
+  flatMap<U>(fn: (entry: T, index: number) => U): GatsbyIterable<U> {
+    return new GatsbyIterable(() => flatMapSequence(this, fn))
+  }
+
+  slice(start: number, end?: number): GatsbyIterable<T> {
+    if ((typeof end !== `undefined` && end < start) || start < 0)
+      throw new Error(
+        `Both arguments must not be negative and end must be greater than start`
+      )
+    return new GatsbyIterable<T>(() => sliceSequence(this, start, end))
+  }
+
+  deduplicate(keyFn?: (entry: T) => unknown): GatsbyIterable<T> {
+    return new GatsbyIterable<T>(() => deduplicateSequence(this, keyFn))
   }
 
-  forEach(callback: (entry: T) => unknown): void {
+  forEach(callback: (entry: T, index: number) => unknown): void {
+    let i = 0
     for (const value of this) {
-      callback(value)
+      callback(value, i++)
     }
   }
+
+  /**
+   * Assuming both this and the other iterable are sorted
+   * produces the new sorted iterable with interleaved values.
+   *
+   * Note: this method is not removing duplicates
+   */
+  mergeSorted<U = T>(
+    other: Iterable<U>,
+    comparator?: (a: T | U, b: T | U) => number
+  ): GatsbyIterable<T | U> {
+    return new GatsbyIterable(() => mergeSorted(this, other, comparator))
+  }
+
+  /**
+   * Assuming both this and the other iterable are sorted
+   * produces the new sorted iterable with values from this iterable
+   * that also exist in the other iterable.
+   */
+  intersectSorted<U = T>(
+    other: Iterable<U>,
+    comparator?: (a: T | U, b: T | U) => number
+  ): GatsbyIterable<T | U> {
+    return new GatsbyIterable(() => intersectSorted(this, other, comparator))
+  }
+
+  /**
+   * Assuming this iterable is sorted, removes duplicates from it
+   * by applying comparator(prev, current) to sibling iterable values.
+   *
+   * Comparator function is expected to return 0 when items are equal,
+   * similar to Array.prototype.sort() argument.
+   *
+   * If comparator is not set, uses strict === comparison
+   */
+  deduplicateSorted(comparator?: (a: T, b: T) => number): GatsbyIterable<T> {
+    return new GatsbyIterable<T>(() => deduplicateSorted(this, comparator))
+  }
```

Comment:
Not sure how this will be used, but this method existing on iterable that we don't know if is sorted seems dangerous (it might be easy to change something not to be sorted anyway and continue using this method without TS being unhappy about it) - would having `GatsbySortedIterable` with this method added feasible just so TS can shield us from something like that? It not being worth the effort also is fine :)

Reply:
(same for `mergeSorted`)

Reply:
Good point. But I afraid TS cannot shield us from creating `GatsbySortedIterable` that is not sorted, right? So it is somewhat always based on assumptions unless we add a runtime check which could be expensive for complex objects.

Reply:
Take for instance `.map` method - should it exist on `GatsbySortedIterable`? And if yes - should it return `GatsbySortedIterable` or `GatsbyIterable`?

Reply:
Yeah, you are right - at one point or another we would have to force TS to treat something as sorted without having any runtime guarantees, so I guess as-is is fine